[DB-40533] Add support for new type VECTOR(<dim>,DOUBLE)

alampe3ds opened the pull request:
Add encode/decode support for new VECTOR datatype.

A result column of VECTOR(<dim>, DOUBLE) shows as type datatype.VECTOR_DOUBLE in the metadata description.

It is returned as datatype.Vector with subtype
datatype.Vector.DOUBLE which can be used as a list.

For a parameter to be detected as VECTOR(<dim>,DOUBLE) it also needs to use the datatype.Vector datatype. Alternatively, binding a string is also possible - which currently includes binding a list of numbers which will then be converted to string.

This commit also adds tests for the described behavior.

alampe3ds commented on tests/nuodb_types_test.py:
```diff
@@ -125,3 +127,87 @@ def test_null_type(self):
+        con = self._connect()
+        cursor = con.cursor()
+        
+        # only activate this tests if tested against version 8 or above
+        cursor.execute("select cast(substring_index(release_ver, '.', 1) as int)"
+                       " from system.nodes limit 1")
+        row = cursor.fetchone()
+        database_major_version = row[0]
+        if database_major_version < 8:
+            return
```
I'll look into that. If it isn't extremely hard (which I doubt) I'll do that.

alampe3ds replied:
Changed that.

alampe3ds commented on tests/nuodb_types_test.py:
```diff
@@ -125,3 +127,87 @@ def test_null_type(self):
+        
+        # only activate this tests if tested against version 8 or above
+        cursor.execute("select cast(substring_index(release_ver, '.', 1) as int)"
+                       " from system.nodes limit 1")
```
Changed to that function now. The version numbering makes it harder to see this is about version 8, but appart from that it is more readable and less brittle that way.

madscientist commented on tests/conftest.py:
```diff
@@ -299,10 +299,21 @@ def database(ap, db, te):
+                try:
+                    cursor.execute("select GETEFFECTIVEPLATFORMVERSION() from system.dual")
+                    row = cursor.fetchone()
+                    effective_version = row[0]
+                    system_information['effective_version'] = effective_version
```
It's a little funny to me to assign a new local variable effective_version, just to then re-assign it again... I spent some seconds looking for where else this new variable was used.  But it works.

alampe3ds replied:
I like that style of coding with names for stuff. But this time I might have gone too far. Will shorten it a bit :-)

madscientist commented on tests/nuodb_base.py:
```diff
@@ -25,17 +25,19 @@ class NuoBase(object):
 
         # Verify the database is up and has a running TE
-        dbname = database['database']
+        dbname = database['connect_args']['database']
```
Why not just use `self.conect_args['database']` here?

alampe3ds replied:
Wanted to make this a simple substitution for review purposes. Will change that, there is no benefit to doing it this way.

madscientist commented on tests/nuodb_types_test.py:
```diff
@@ -131,13 +131,9 @@ def test_null_type(self):
-        row = cursor.fetchone()
-        database_major_version = row[0]
-        if database_major_version < 8:
+        if self.system_information['effective_version'] < 1835008:
```
I looked and I don't see a new Version.cfg feature defined for VECTOR... is it really true that we've so far implemented this facility without introducing any (database) protocol changes at all?  That's impressive.

alampe3ds replied:
We talked about that I think during the review for the C++ driver. It is still somewhat open whether we should do a protocol change or not.
Without the change drivers old drivers don't like the new type in a way that they will give somewhat internal decode errors when they get a resultset with them.
With a change, however, the only thing we can really do is throwing a clearer error on the database side during encoding. We discussed about sending a string representation instead, but you had good arguments that this might not be ideal behavior.

There will be a follow-up discussion to that, but it will most likely not affect the drivers.

Do you object to keeping the magic number here? I could add it to a variable NUDB_V8 or something like that to make it less magic.

madscientist replied:
Yes, I was specifically talking about the database protocol, not the client protocol.  No, keep it as-is.  I don't want to create an "8.0" variable.  If there was a feature name we could use, that would be appropriate.

madscientist reviewed (approved):
There are a few small cleanups you can implement, or not, as you like. Thanks for this work!!

butson reviewed (commented):
```
datatype.__all__ should include Vector.
```
 As user of pynuodb would need/want to use pynuodb.Vector not pynuodb.datatype.Vector

alampe3ds commented:
> ```
> datatype.__all__ should include Vector.
> ```
> 
> As user of pynuodb would need/want to use pynuodb.Vector not pynuodb.datatype.Vector

Good point.
I changed that and modified the test to `import pynuodb` instead of the datatype to reflect customer usage there.

alampe3ds commented:
@madscientist Thank you for reminding me. Yes, the branch is merged into master as b9872c5bbc7d6dea41e4b16aec88f6358ddbeb7e - closing the PR now.
